File: templates/javascript/v1/designtemplates/chain/handlers.js
const NOT_IMPLEMENTED_ERROR = "Not implemented";
const TOKEN_PREFIX = "token:";
const UNHANDLED_PREFIX = "Unhandled request: ";
const VALIDATION_FAILED_PREFIX = "Validation failed for request: ";
const AUTH_FAILED_PREFIX = "Auth failed for request: ";
const PROCESSED_PREFIX = "Processed request ";
const PAYLOAD_LENGTH_LABEL = " with payload length ";

function isDefined(value) {
  return value !== null && value !== undefined;
}

function formatUnhandledRequest(requestId) {
  return `${UNHANDLED_PREFIX}${requestId}`;
}

function formatValidationFailure(requestId) {
  return `${VALIDATION_FAILED_PREFIX}${requestId}`;
}

function formatAuthFailure(requestId) {
  return `${AUTH_FAILED_PREFIX}${requestId}`;
}

function formatProcessed(requestId, payloadLength) {
  return `${PROCESSED_PREFIX}${requestId}${PAYLOAD_LENGTH_LABEL}${payloadLength}`;
}

export class Handler {
  constructor() {
    this.next = null;
  }

  linkWith(next) {
    this.next = next;
    return next;
  }

  handle(request) {
    const result = this.process(request);
    if (isDefined(result)) {
      return result;
    }
    if (!this.next) {
      return formatUnhandledRequest(request.id);
    }
    return this.next.handle(request);
  }

  process(_request) {
    throw new Error(NOT_IMPLEMENTED_ERROR);
  }
}

export class ValidationHandler extends Handler {
  process(request) {
    if (!request.payload) {
      return formatValidationFailure(request.id);
    }
    return null;
  }
}

export class AuthHandler extends Handler {
  process(request) {
    if (request.payload.includes(TOKEN_PREFIX)) {
      return null;
    }
    return formatAuthFailure(request.id);
  }
}

export class BusinessHandler extends Handler {
  process(request) {
    return formatProcessed(request.id, request.payload.length);
  }
}

File: templates/python/v1/designtemplates/chain/handlers.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

TOKEN_PREFIX = "token:"
UNHANDLED_PREFIX = "Unhandled request: "
VALIDATION_FAILED_PREFIX = "Validation failed for request: "
AUTH_FAILED_PREFIX = "Auth failed for request: "
PROCESSED_PREFIX = "Processed request "
PAYLOAD_LENGTH_LABEL = " with payload length "


@dataclass(slots=True)
class Request:
    id: str
    payload: str


@dataclass(slots=True)
class Handler:
    next: Optional["Handler"] = None

    def link_with(self, next_handler: "Handler") -> "Handler":
        self.next = next_handler
        return next_handler

    def handle(self, request: Request) -> str:
        result = self.process(request)
        if result is not None:
            return result
        if self.next is None:
            return _format_unhandled(request.id)
        return self.next.handle(request)

    def process(self, request: Request) -> Optional[str]:
        raise NotImplementedError


class ValidationHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        if not request.payload:
            return _format_validation_failed(request.id)
        return None


class AuthHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        if TOKEN_PREFIX in request.payload:
            return None
        return _format_auth_failed(request.id)


class BusinessHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        return _format_processed(request.id, len(request.payload))


def _format_unhandled(request_id: str) -> str:
    return f"{UNHANDLED_PREFIX}{request_id}"


def _format_validation_failed(request_id: str) -> str:
    return f"{VALIDATION_FAILED_PREFIX}{request_id}"


def _format_auth_failed(request_id: str) -> str:
    return f"{AUTH_FAILED_PREFIX}{request_id}"


def _format_processed(request_id: str, payload_length: int) -> str:
    return f"{PROCESSED_PREFIX}{request_id}{PAYLOAD_LENGTH_LABEL}{payload_length}"
